service/storage_service.cc

```diff
@@ -5087,13 +5087,14 @@ future<> storage_service::stream_tablet(locator::global_tablet_id tablet) {
 
         auto& table = _db.local().find_column_family(tablet.table);
         std::vector<sstring> tables = {table.schema()->cf_name()};
-        auto streamer = make_lw_shared<dht::range_streamer>(_db, _stream_manager, tm,
+        auto streamer = make_lw_shared<dht::range_streamer>(_db, _stream_manager, std::move(tm),
                                                             guard.get_abort_source(),
                                                             tm->get_my_id(), _snitch.local()->get_location(),
                                                             format("Tablet {}", trinfo->transition),
                                                             reason,
                                                             topo_guard,
                                                             std::move(tables));
+        tm = nullptr;
         streamer->add_source_filter(std::make_unique<dht::range_streamer::failure_detector_source_filter>(
                 _gossiper.get_unreachable_members()));
 
```

Comment:
we can no longer rely on trinfo below when sync'ing migration and split, right?

suggest to add this change:

```
diff --git a/service/storage_service.cc b/service/storage_service.cc
index 4f1d178379..89f6a623f5 100644
--- a/service/storage_service.cc
+++ b/service/storage_service.cc
@@ -5103,11 +5103,12 @@ future<> storage_service::stream_tablet(locator::global_tablet_id tablet) {
         auto& tinfo = tmap.get_tablet_info(tablet.tablet);
         auto range = tmap.get_token_range(tablet.tablet);
         locator::tablet_replica leaving_replica = locator::get_leaving_replica(tinfo, *trinfo);
+        locator::tablet_replica pending_replica = trinfo->pending_replica;
         if (leaving_replica.host == tm->get_my_id()) {
             // The algorithm doesn't work with tablet migration within the same node because
             // it assumes there is only one tablet replica, picked by the sharder, on local node.
             throw std::runtime_error(format("Cannot stream within the same node, tablet: {}, shard {} -> {}",
-                                            tablet, leaving_replica.shard, trinfo->pending_replica.shard));
+                                            tablet, leaving_replica.shard, pending_replica.shard));
         }
 
         locator::tablet_migration_streaming_info streaming_info = get_migration_streaming_info(tm->get_topology(), tinfo, *trinfo);
@@ -5186,7 +5187,7 @@ future<> storage_service::stream_tablet(locator::global_tablet_id tablet) {
         //  participating in streaming anymore (which is true), so it could schedule more
         //  migrations. This way compaction would run in parallel with streaming which can
         //  reduce the delay.
-        co_await _db.invoke_on(trinfo->pending_replica.shard, [tablet] (replica::database& db) {
+        co_await _db.invoke_on(pending_replica.shard, [tablet] (replica::database& db) {
             auto& table = db.find_column_family(tablet.table);
             return table.maybe_split_compaction_group_of(tablet.tablet);
         });
```

Reply:
Right